### PR TITLE
feat(lsp): add internal debugging logging

### PR DESCRIPTION
**Pull request by kitsonk**
Ref: #10368

The lsp would respect a setting which will log all the marks and measures to stderr along with their arguments which can be used to investigate certain issues (as well as more easily generate test cases).  This requires a client which can pass the configuration option (see: denoland/vscode_deno#406).

### CODE REVIEW

File: cli/lsp/language_server.rs [hunk 1]
```diff
@@ -343,8 +351,14 @@ impl Inner {
     Ok(())
   }
 
+  pub fn update_performance(&mut self) {
+    self
+      .performance
+      .do_logging(self.config.settings.internal_debug);
+  }
```
Comment (ry) [inline cli/lsp/language_server.rs:358]: We already have a system for conditional logging using the `debug!` and `info!` macros. These can be enabled with the `--log-level=debug` command-line option. Can we not reuse that system for LSP debugging?

Note that the log-level could be adjusted at runtime as well: 
https://docs.rs/log/0.4.14/log/#implementing-a-logger

File: cli/lsp/performance.rs [hunk 2]
```diff
@@ -129,14 +133,38 @@ impl Performance {
...
+  pub fn mark<S: AsRef<str>, V: Serialize>(
+    &self,
+    name: S,
+    maybe_args: Option<V>,
+  ) -> PerformanceMark {
     let name = name.as_ref();
     let mut counts = self.counts.lock().unwrap();
     let count = counts.entry(name.to_string()).or_insert(0);
     *count += 1;
+    if self.logging_enabled {
```
Comment (ry) [inline cli/lsp/performance.rs:152]: Ideally this should use the conditional logging provided by the rust log / env_logger crate.

File: cli/lsp/language_server.rs (hunk 1 above)
Comment (kitsonk) [reply cli/lsp/language_server.rs:358]: Yeah, I knew that, the problem is a lot of other crates, especially swc, use the same logging, and so setting the logging flag makes it impossible to sort out what is going on.

Comment (ry) [reply cli/lsp/language_server.rs:358]: env_logger provides the ability to filter messages based on regex
https://docs.rs/env_logger/0.8.3/env_logger/#filtering-results

Comment (kitsonk) [reply cli/lsp/language_server.rs:358]: Ok, I spent some time looking at this.  A few notes:

- the logger can only be set once and has to be global static
- we use env_logger as our global logger to do this, once we `init()` it appears you can't reconfigure it.
- the `RUST_LOG` is an environment variable, which can be passed in at startup, and the filter `debug/mark|measure` would be suitable, but we can only do that when start/restarting the LSP.
- you can use the `.parse_filters()` method on the builder which follows the same format as `RUST_LOG` but again that has to be determined at startup, versus changing dynamically.

There is the `.format()` closure which maybe could do some "filtering" of records based on a flag, but there doesn't appear to be a dynamic way of changing things in a running process, outside of implementing our own logger (which is what is mentioned in the first link).

Comment (ry) [reply cli/lsp/language_server.rs:358]: My guess is the Log::enabled method could be used to dynamically turn on and off logging
https://docs.rs/log/0.4.14/log/trait.Log.html#tymethod.enabled

But it seems a little difficult to access... I guess you'd have to wrap the env_logger.

Comment (kitsonk) [reply cli/lsp/language_server.rs:358]: Hmmm... yeah, that should work, but like you said would require wrapping. Ok, I will look at that a bit closer. You can generate the logger without installing it for wrapping, that I know.

Comment (kitsonk) [reply cli/lsp/language_server.rs:371]: Ok, I reworked it... the biggest "problem" is that we need to share the state of the logging being enabled between the logger and the `lsp::language_server` and we don't have any infrastructure for that, so I chose a way (`Arc<AtomicBool>`) to do that, but am more than willing to take other suggestions.

Also broke all the logging stuff out to a seperate module, as it just seem unnecessary in an already complex `main.rs`.

File: cli/main.rs [hunk 3]
```diff
@@ -1147,7 +1111,12 @@ pub fn main() {
   if !flags.v8_flags.is_empty() {
     init_v8_flags(&*flags.v8_flags);
   }
-  init_logger(flags.log_level);
 
-  unwrap_or_exit(tokio_util::run_basic(get_subcommand(flags)));
+  // when this flag is set to `true`, the process will log specific diagnostic
+  // debug information,
+  let lsp_debug_flag = Arc::new(AtomicBool::new(false));
```
Comment (ry) [inline cli/main.rs:1117]: It's unfortunate that this `lsp_debug_flag` has leaked all the way to top-level of the program and many signatures have to be modified to pass it around.

What if you use lazy_static to have a global `Arc<AtomicBool>` in `cli/logger.rs`? Then `cli/lsp/language_server.rs` could reference that directly without having to pass it around like this. I think practically this `lsp_debug_flag` is global anyway.

Comment (kitsonk) [reply cli/main.rs:1117]: Yeah, that makes more sense. I will try that.

**Review by ry — APPROVED** [review]
LGTM

**Pull request merged**